# About this tutorial

## Overview

We will consider the basic informatics problem of _sequence alignment_ from a
practical perspective.

Our goal is to consider _sequence alignment_ from different angles, and
meanwhile develop algorithmic thinking, introduce general concepts for
informatics,

## Contents

Techniques: Brute force, Dynamic programming, Divide-and-Conquer, Greedy
Combinatorics,
Graphs (Shortest paths, Spanning trees), Strings, Hashing

Performance: Problem complexity, Absolute performance and scaling, Memory and
runtime, Worst-case and average-case asymptotics, Amortized analysis

Correctness: Invariants, Approximate vs Exact, Heuristic, Evaluation

Central problem: Sequence alignment
Pangenomics
Data: Sequencing data
Applications: Each algorithm will be applied to a basic biological problem
Programming: You will be asked to submit your solutions on rosalind.info
Heuristic: Genetic algorithms, research tasks on request
Central problems: Sequence alignment (global, semi-global, local, suboptimal, MSA), Assembly, Pangenomics
Algorithms: Exact, approximate, probabilistic, heuristic algorithms; Greedy, Minimizers
Data structures: Genome graph, Suffix tree and suffix array, Bloom filter, HMMs
Optimizations: Four Russians method, Bit-parallel
Tools: State-of-the-art aligners (minimap2, Edlib, WFA, AStarix, A*PA, GraphAligner)
Homeworks: rosalind.info problems, edit a Wikipedia page in computational biology

## Goal

Olympiad informatics, 

## Prerequisites 

You will need to feel confortable with middle-school mathematics and have basic
programming skillsw with a language like C++ or Python.
